server/api.go

```diff
@@ -95,13 +101,34 @@ func (p *Plugin) connectUserToGitHub(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
-	conf := p.getOAuthConfig()
+	privateAllowed := false
+	pVal := r.URL.Query().Get("private")
+	if pVal == "ok" {
+		privateAllowed = true
+	}
 
-	state := fmt.Sprintf("%v_%v", model.NewId()[0:15], userID)
+	conf := p.getOAuthConfig(privateAllowed)
 
-	p.API.KVSet(state, []byte(state))
+	state := OAuthState{
+		UserID:         userID,
+		Token:          model.NewId()[:15],
+		PrivateAllowed: privateAllowed,
+	}
 
-	url := conf.AuthCodeURL(state, oauth2.AccessTypeOffline)
+	stateBytes, err := json.Marshal(state)
+	if err != nil {
+		http.Error(w, "json marshal failed", http.StatusInternalServerError)
+		return
+	}
+
+	appErr := p.API.KVSet(state.Token, stateBytes)
```

Comment:
I know this already existed, but this surprises me: I would have expected the flow to be stateless, with the full state being returned back to the plugin signed so as to be trusted.

Reply:
As another side, should we `KVSetWithExpiry` to ensure these key values are cleaned up even if the oAuth flow never terminates?

Reply:
KVSetWithExpiry is a great idea (I should have thought of that) -- added.

This is the way we did it in Jira, and the way the Github plugin was doing it -- but regardless of "how we've always done it" (a bad argument by itself, imho), I think storing the one-time token is the right way to do it. From the oauth2 docs: https://godoc.org/golang.org/x/oauth2#Config.AuthCodeURL 

```
State is a token to protect the user from CSRF attacks. You must always provide a non-empty string and validate that it matches the state query parameter on your redirect callback. See http://tools.ietf.org/html/rfc6749#section-10.12 for more info.
```

If that's not what we're doing here, _please_ let me know (because I've been doing it wrong).

Reply:
I think your approach is correct -- I was vaguely recalling a SAML implementation from a lifetime ago, but realized that the key value store here replaces the cookies that otherwise might be used to do the same thing. Thanks for the link and the chance to learn something :)